extensions/amp-pinterest/0.1/amp-pinterest.js

```diff
@@ -0,0 +1,574 @@
+  'menubar=no,width=900,height=500,left=0,top=0';
+
+const BASE60 = '0123456789ABCDEFGHJKLMNPQRSTUVWXYZ_abcdefghijkmnopqrstuvwxyz';
+
+class AmpPinterest extends AMP.BaseElement {
+  /** @override */
+  isLayoutSupported(layout) {
+    return isLayoutSizeDefined(layout);
+  }
+  /** @override */
+  layoutCallback() {
+
+    let height = this.element.getAttribute('height');
+
+    let width = this.element.getAttribute('width');
+
+    let pinDo = AMP.assert(this.element.getAttribute('data-do'),
+      'The data-do attribute is required for <amp-pinterest> %s',
+      this.element);
+
+    let pinUrl = AMP.assert(this.element.getAttribute('data-url'),
+      'The data-url attribute is required for <amp-pinterest> %s',
+      this.element);
+
+    // make a 12-digit base-60 number for performance tracking
+    let guid = '';
+    for (let i = 0; i < 12; i = i + 1) {
+      guid = guid + BASE60.substr(Math.floor(Math.random() * 60), 1);
+    }
+
+    // open a new window
+    let pop = function (url, shouldPop) {
+      if (shouldPop) {
+        // amp=1&guid=guid are already in the query before long fields
+        window.open(url, '_pinit', POP);
+      } else {
+        // this is a straight window.open; append amp=1&guid=guid
+        window.open(url + '?amp=1&guid=' + guid, '_blank');
+      }
+    };
+
+    // log a string
+    let log = function (str) {
+      let call = new Image();
+      let query = 'https://log.pinterest.com/?guid=' + guid;
+      query = query + '&amp=1';
+      // add optional string &foo=bar
+      if (str) {
+        query = query + str;
+      }
+      query = query + '&via=' + document.URL;
+      call.src = query;
+    };
+
+    // load data from API via XHR
+    let call = function(url) {
+      return new Promise((resolve, reject) => {
+        let xhr = new XMLHttpRequest();
+        xhr.onreadystatechange = function() {
+          if (xhr.readyState == 4) {
+            if (xhr.status == 200) {
+              // all replies are JSON and should parse
+              try {
+                let data = JSON.parse(xhr.responseText);
+                resolve(data);
+              } catch (err) {
+                reject(new Error('API reply was not JSON: ' + url));
+              }
+            } else {
+              reject(new Error('API call failed: ' + url));
+            }
+          }
+        };
+        xhr.open("GET", url, true);
+        xhr.send();
+      });
+    };
+
+    // set a DOM property or text attribute
+    let set = function (el, att, string) {
+      if (typeof el[att] === 'string') {
+        el[att] = string;
+      } else {
+        el.setAttribute(att, string);
+      }
+    };
+
+    // create a DOM element
+    let make = function (obj) {
+      let el = false, tag, att;
+      for (tag in obj) {
+        el = document.createElement(tag);
+        for (att in obj[tag]) {
+          if (typeof obj[tag][att] === 'string') {
+            set(el, att, obj[tag][att]);
```

Comment:
are we just propagating dom element attributes here? would [`this.propagateAttributes(attributes, element)`](https://github.com/ampproject/amphtml/blob/master/src/base-element.js#L327) found in `BaseElement` work somewhere here?